src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/Utf8StringMarshaller.cs

```diff
@@ -37,32 +37,33 @@ public Utf8StringMarshaller(string? str)
         /// </remarks>
         public Utf8StringMarshaller(string? str, Span<byte> buffer)
         {
-            _allocated = null;
+            _allocated = false;
+
             if (str is null)
             {
-                _span = default;
+                _nativeValue = null;
                 return;
             }
 
+            const int MaxUtf8BytesPerChar = 3;
+
             // + 1 for null terminator
-            int maxByteCount =  Encoding.UTF8.GetMaxByteCount(str.Length) + 1;
-            if (buffer.Length >= maxByteCount)
-            {
-                int byteCount = Encoding.UTF8.GetBytes(str, buffer);
-                buffer[byteCount] = 0; // null-terminate
-                _span = buffer;
-            }
-            else
+            int maxByteCount = checked(MaxUtf8BytesPerChar * str.Length + 1);
+            if (maxByteCount > buffer.Length)
             {
-                _allocated = (byte*)Marshal.AllocCoTaskMem(maxByteCount);
-                int byteCount;
-                fixed (char* ptr = str)
+                // Calculate accurate byte count when the provided stack-allocated buffer is not sufficient
+                maxByteCount = Encoding.UTF8.GetByteCount(str) + 1;
+                if (maxByteCount > buffer.Length)
```

Comment:
This makes the marshaller faster for ascii strings of lengths between 85 and 256. The cost of calculating the exact buffer length is less than the cost of going through the Alloc/Free path.